java/ql/src/Security/CWE/CWE-926/ContentProviderIncompletePermissions.qhelp

```diff
@@ -0,0 +1,56 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+<p>In the Android manifest file, an application's <code>provider</code> elements
+  define the permissions necessary to access a resource using that provider.
+  Permissions are specified with
+  the <code>android:readPermission</code>, <code>android:writePermission</code>,
+  or <code>android:permission</code> attributes. If an application only
+  specifies the <code>android:readPermission</code>
+  or <code>android:writePermission</code> attribute, no permissions will be
+  required to do other operations.
```

Comment:
I'm not entirely sure if I've understood this section correctly. It sounds as if we're saying that by default, an application has permission to do anything via a provider, unless the permissions are explicitly defined. So you need to specify the permissions required for each operation.

Based on that understanding, maybe this would work?

```suggestion
<p>The Android manifest file contains <code>provider</code> elements that you should use to specify the explicit permissions an application requires to access a resource using that provider. 
  You specify the permissions using
  the <code>android:readPermission</code>, <code>android:writePermission</code>,
  or <code>android:permission</code> attributes. 
If you do not specify the permission required to perform an operation, the application will implicitly have access to perform that operation.  
  For example, if you specify only <code>android:readPermission</code>, the application must have explicit permission to read data, but requires no permission to write data.
```

Reply:
@felicitymay Thank you for the review! Most of the suggested edit sounds good. But looking at this again, I'm not sure either first sentence is saying quite what I want to say. 

Applications have [content providers], which define how other applications can access the content an application makes available. The manifest file encodes which providers are accessible to other apps using the `<provider>` element. The provider element has permission attributes which specify which operations can be performed.

So maybe something more like:

> The Android manifest file specifies the content providers for the application using `<provider>` elements. The `<provider>` element specifies the explicit permissions an application requires in order to access a resource using that provider.

How does that sound?

[content providers]: https://developer.android.com/reference/android/content/ContentProvider

Reply:
Thank you. That sounds good to me.